dask_expr/_merge.py

```diff
@@ -213,6 +238,96 @@ def _simplify_up(self, parent):
                 return result[parent_columns]
 
 
+class HashJoinP2P(Merge, PartitionsFiltered):
+    _parameters = [
+        "left",
+        "right",
+        "how",
+        "left_on",
+        "right_on",
+        "suffixes",
+        "indicator",
+        "_partitions",
+    ]
+    _defaults = {
+        "how": "inner",
+        "left_on": None,
+        "right_on": None,
+        "suffixes": ("_x", "_y"),
+        "indicator": False,
+        "_partitions": None,
+    }
+
+    def _lower(self):
+        return None
+
+    @functools.cached_property
+    def _meta(self):
+        left = self.left._meta.drop(columns=_HASH_COLUMN_NAME)
+        right = self.right._meta.drop(columns=_HASH_COLUMN_NAME)
+        return left.merge(
+            right,
+            left_on=self.left_on,
+            right_on=self.right_on,
+            indicator=self.indicator,
+            suffixes=self.suffixes,
+        )
+
+    def _layer(self) -> dict:
+        dsk = {}
+        name_left = "hash-join-transfer-" + self.left._name
+        name_right = "hash-join-transfer-" + self.right._name
+        transfer_keys_left = list()
+        transfer_keys_right = list()
+        for i in range(self.left.npartitions):
+            transfer_keys_left.append((name_left, i))
+            dsk[(name_left, i)] = (
+                merge_transfer,
+                (self.left._name, i),
+                self.left._name,
+                i,
+                self.npartitions,
+                self._partitions,
+            )
+        for i in range(self.right.npartitions):
+            transfer_keys_right.append((name_right, i))
+            dsk[(name_right, i)] = (
+                merge_transfer,
+                (self.right._name, i),
+                self.right._name,
+                i,
+                self.npartitions,
+                self._partitions,
+            )
+
+        _barrier_key_left = barrier_key(ShuffleId(self.left._name))
+        _barrier_key_right = barrier_key(ShuffleId(self.right._name))
+        dsk[_barrier_key_left] = (shuffle_barrier, self.left._name, transfer_keys_left)
+        dsk[_barrier_key_right] = (
+            shuffle_barrier,
+            self.right._name,
+            transfer_keys_right,
+        )
+
+        for part_out in self._partitions:
+            dsk[(self._name, part_out)] = (
+                merge_unpack,
+                self.left._name,
+                self.right._name,
+                part_out,
+                _barrier_key_left,
+                _barrier_key_right,
+                self.how,
+                self.left_on,
+                self.right_on,
+                self.left._meta.drop(columns=_HASH_COLUMN_NAME),
```

Comment:
I don't love how the p2p merge depends on `_HASH_COLUMN_NAME` and drops this column, but isn't responsible for generating this column. Unfortunately, the main issue is that `merge_unpack` drops the shuffle column, so there is not much we can do here to make the algorithm more intuitive.

To clarify, it would be much clearer if we could just lower the merge to "add partitioning columns" + "p2p merge" + "drop partitioning column"

Reply:
Yeah I mostly agree but I fixed something else there today. This is on my medium term todo list